cranelift/isle/isle/src/overlap.rs

```diff
@@ -0,0 +1,876 @@
+//! Overlap detection for rules in ISLE.
+
+use crate::error::{Error, Result, Source, Span};
+use crate::sema::{
+    self, ConstructorKind, Rule, RuleId, Sym, Term, TermEnv, TermId, TermKind, Type, TypeEnv,
+    TypeId, VarId,
+};
+
+/// Check for overlap.
+pub fn check(tyenv: &TypeEnv, termenv: &TermEnv) -> Result<()> {
+    let env = Env::new(tyenv, termenv);
+    let mut errors = Errors::new(termenv.rules.len());
+    for term in termenv.terms.iter() {
+        // The only isle declaration that currently produces overlap is constructors whose
+        // definition is entirely in isle.
+        if !env.is_internal_constructor(term.id) {
+            continue;
+        }
+
+        check_overlap_groups(&mut errors, &env, term);
+    }
+
+    if !errors.is_empty() {
+        let mut errors = errors.report(&env);
+        return match errors.len() {
+            1 => Err(errors.pop().unwrap()),
+            _ => Err(Error::Errors(std::mem::take(&mut errors))),
+        };
+    }
+
+    Ok(())
+}
+
+/// A node in the error graph.
+struct Node {
+    /// The number of other rules this node overlaps with.
+    degree: usize,
+
+    /// `true` entries where an edge exists to the node at that index.
+    edges: Vec<bool>,
+}
+
+impl Node {
+    /// Make a new `Node` in the error graph with the given number of total nodes.
+    fn new(len: usize) -> Self {
+        let mut edges = Vec::with_capacity(len);
+        edges.resize(len, false);
+        Self { degree: 0, edges }
+    }
+
+    /// Add an edge between this node and the other node.
+    fn add_edge(&mut self, other: RuleId) {
+        if self.edges[other.0] {
+            return;
+        }
+
+        self.degree += 1;
+        self.edges[other.0] = true;
+    }
+
+    /// Remove an edge between this node and another node.
+    fn remove_edge(&mut self, other: RuleId) {
+        if !self.edges[other.0] {
+            return;
+        }
+
+        self.degree -= 1;
+        self.edges[other.0] = false;
+    }
+}
+
+/// A graph of all the rules in the isle source, with bi-directional edges between rules that are
+/// discovered to have overlap problems.
+struct Errors {
+    /// Edges between rules indicating overlap. As the edges are not directed, the edges are
+    /// normalized by ordering the rule ids.
+    nodes: Vec<Node>,
```

Comment:
Is the comment that "the edges are normalized by ordering the rule ids" still true? I'm guessing that's left over from an earlier version.

Like for `Node`, I'd prefer `HashMap` over `Vec` so there's no heap allocation unless an error is actually found. And like for `Node` that means you can use `#[derive(Default)]`.